### ISSUE TITLE
Startup dependency sequencing issues

### ISSUE THREAD

**Issue by BobDickinson** [joyent/containerpilot#273]
I have successfully implemented a relatively sophisticated AutoPilot solution for our Synchro product (which incorporates our own Node.js server, StashBox, Nginx, Redis, Consul, and Prometheus).  I'm also working on a Parse solution that incorporates the Parse Node.js app, Nginx, Redis, Mongodb, Consul, and Prometheus.  I have a kind of recurring problem that makes me think maybe I'm missing something about how ContainerPilot solutions are supposed to work.

My issue tends to be around services that take a while to get into a usable operating state.  Consul, Mongodb, and Redis all have this very entertaining behavior where they go through a whole byzantine process on startup where they see they are new, they wonder who their leader is, they decide to have an election, they see who wants to vote, it's just them voting, they vote, WINNER, they are the master.  This takes a significant (but variable) amount of time, during which the services are practically unavailable to other dependent services (the services are running, presumably registered with Consul if Consul is up, and might even pass a health check, but they are not usable yet).

For example, my own Node.js server cannot start until Redis is up and running (and has determined that it is the master and is available for write).  It's actually more complicated than that, as my Redis support uses Sentinel config, which I have to get from Consul (so consul also has to have finished his democratic process of determining that the sole instance is actually in charge).  In my other solution, the Parse server cannot start until Mongodb is up and running (and has determined that it is the master and is available for write).

So in practice what happens is a lot of smashing and grinding of gears on startup.  In my own app, I just added a configurable timeout and retry count so that my app would wait until Consul and Redis were available before it continued startup.  But in my Parse case, that's not my code.  It starts up, tries to contact Mongodb, fails (because the sole instance of Mongodb still hasn't determined that it is the master), and that's it.  No joy.

Am I doing something philosophically wrong here?  Is there a way I can make the dependent services delay their startup via ContainerPilot until the services they depend on are actually available?  If I specify a backend in my containerpilot.json, does ContainerPilot make sure that that backend is available before starting my service?  That the backend has passed a health check?  If I had a health check that made sure this whole election / master situation was resolved before it reported healthy, would that make it any better?

**Comment by Lennie** [comment]
Thank you for posting this. I've also been looking at using Container Pilot in projects and thinking of similar problems and how to solve them.

The last few months I've not been following Container Pilot as closely as I've wanted, so maybe this was already done/better/differently to solve it.

One thing I was thinking is to have one or more extra start up states. During one of those states the presence of the service for which Container Pilot is running could be announced to Consul for peers to find each other, but not ready to be used by other services.

If I look at for example https://github.com/zalando/patroni it looked to me like it has multiple phases to how it starts up as well.

**Comment by tgross** [comment]
> That the backend has passed a health check? If I had a health check that made sure this whole election / master situation was resolved before it reported healthy, would that make it any better?

This is exactly the right solution. A service shouldn't pass health checks until it's ready to work.

ContainerPilot will only report backends that are marked healthy. So "healthy" means whatever you need it to mean in the context of your application. You can also use `preStart` to fetch configuration data you might need from Consul in cases where an application doesn't allow for config updates.

So in your examples a startup sequence might go something like:
- Redis gets Sentinel configuration from Consul in `preStart`, polling Consul until ready.
- Redis instances start and register themselves as primary vs replica via ContainerPilot `health` check.
- Node.js server tries to get the Mongo config via its `backend.onChange`, but there are no backends yet so it doesn't pass health checks yet.
- Parse server can't start safely without Mongo, so poll Consul in `preStart` instead
- Mongo starts up
- Parse server's `preStart` sees Mongo and continues on to starting the application.
- Node.js server receives `backend.onChange` and gracefully reloads, allowing the `health` checks to pass.
- etc.

So you have to handle the dependency chain via the various handlers and allow the stack to "converge" to a ready state on its own, but in exchange you don't need an external service that enforces the startup order and externally asserts their readiness. In practice this means the initial startup of a stack takes just about as long but that it's able to automatically respond to changes by converging to a new ready state.

We haven't eliminated the complexity of dependency management -- that doesn't work without proscribing a narrow set of behaviors like 12Factor, which rules out stateful applications -- but we have tamed that complexity by moving it into the container so that it's owned by the people who know the most about what the application needs. As you can see from your Node.js vs. Parse example, some services need more care and feeding!

**Comment by BobDickinson** [comment]
OK, so just to make sure I'm following this.  In my service requiring that a dependency be healthy before starting, I check consul for the dependency in my preStart and wait there until ready.  I'm assuming something like the top of: 

https://github.com/autopilotpattern/wordpress/blob/master/bin/prestart.sh

If I'm using consul agent (which seems to be current best practices), I was under the impression that it will not be running yet in my preStart and thus I won't have access to consul there (I think the Faithlife Redis example moved all of their preStart logic into an onStart handler due to this).  Is that the case?  Is moving this "wait for healthy dependencies" logic into an onStart (basically into the shell script that actually starts the app) the correct solution to this when using consul agent to access consul?

**Comment by misterbisson** [comment]
I'm late to join this conversation, but when I saw the OP I wanted to share the following diagram to help contextualize the `preStart` hooks:

<img width="906" alt="screen shot 2017-01-25 at 9 19 57 am" src="https://cloud.githubusercontent.com/assets/1709697/22304246/82d7fb48-e2eb-11e6-9e8d-6be009592189.png">

It's already been discussed how that and the health checks can be used to test for dependencies, so I'll just leave that here for others who find the conversation later.

**Comment by tgross** [comment]
> Is moving this "wait for healthy dependencies" logic into an onStart (basically into the shell script that actually starts the app) the correct solution to this when using consul agent to access consul?

You can also do it in the health check itself, as we've done in `autopilotpattern/mysql` and couple other places. This has the added benefit that if the dependency goes away in a transient way (ex. a netsplit) the application now has the opportunity to do something about it like mark itself unhealthy or change its behavior.

**Comment by BobDickinson** [comment]
I understand one of the issues, which is that the dependency service (in my case, Mongodb), needs to report healthy when it is usable by the depending service.  In the `autopilotpattern/mongodb` example that I based my implementation on, someone decided that "health" would be defined by whether you could connect to Mongo (and not whether it was actually ready to "do database stuff").  Looking at the health check here:

https://github.com/autopilotpattern/mongodb/blob/master/bin/manage.py

that change seems pretty straightforward (I only want to return true when it's a running master or slave, and false in all other cases).

The other half of the problem is having the depending service wait until the dependency is healthy before starting.  Your last comment in that area confused me somewhat.  In the depending service (Parse in this case), I want to wait until Mongodb is up before I start.  I can't do this in preStart if I want to use consul agent (which seems to be the current recommendation), so I was presuming to wrap the process start in an "onStart" shell script (where I would have access to consul via consul agent) and do the wait there.  Are you saying I should instead (or also) do that check in the Parse health check?  I didn't see anything like that (depending service evaluating dependency service in its health check) in the example you cited.

**Comment by misterbisson** [comment]
> "health" would be defined by whether you could connect to Mongo (and not whether it was actually ready to "do database stuff")

Would you mind opening a ticket in https://github.com/autopilotpattern/mongodb if you don't already have a PR for that in mind?

**Comment by BobDickinson** [comment]
> Would you mind opening a ticket in https://github.com/autopilotpattern/mongodb if you don't already have a PR for that in mind?

Sure.  Was planning on a PR once I got everything working.

**Comment by misterbisson** [comment]
> If I'm using consul agent (which seems to be current best practices), I was under the impression that it will not be running yet in my preStart and thus I won't have access to consul there

@tgross and I had talked about a different behavior there because of use-cases like this. Before I get into that, [Tim's suggestion from above](https://github.com/joyent/containerpilot/issues/273#issuecomment-275232669) is a good one:

> You can also do it in the health check itself, as we've done in autopilotpattern/mysql and couple other places. 

There are added benefits as Tim pointed out there, but it _won't_ work if your app just won't start without the backends configured.

We have that problem with Nginx. In that case, the `preStart` script connects to the Consul cluster directly, without going through the agent:

https://github.com/autopilotpattern/nginx/blob/master/bin/reload.sh#L4

In the bigger picture, I have to admit I'm struggling to remember some conversations @tim and I had about this in the context of https://github.com/joyent/containerpilot/issues/246.

**Comment by BobDickinson** [comment]
> but it won't work if your app just won't start without the backends configured

I could just go in and fix Parse (I did exactly that to my own app in a similar situation with Redis), but I'm trying to approach this from the standpoint that we should be able to deal with this all externally via AutoPilot/ContainerPilot.

>  In that case, the preStart script connects to the Consul cluster directly, without going through the agent

OK, I think pending a fix to #246, I'll do that.  Thanks.
